README.rst
.. image:: https://git.vaultit.org/common/qvarn-mr/badges/master/build.svg
   :target: https://git.vaultit.org/common/qvarn-mr/commits/master

.. image:: https://git.vaultit.org/common/qvarn-mr/badges/master/coverage.svg
   :target: https://git.vaultit.org/common/qvarn-mr/commits/master


Experimental Qvarn map/reduce service
#####################################

The idea is to create a service, that will listen for changes on specified
resource types. All changes will be routed through defined map/reduce handlers.
And the output will be written into derived resource types.

This service should allow to join several resources into one resource type and
do aggregations via reduce functions.

You can read more about map/reduce in the Wikipedia_.

.. _Wikipedia: https://en.wikipedia.org/wiki/MapReduce


Status
======

The service is fully implemented and has near 100% test coverage, but is not
yet used by any real project. So theoretically it should work, but practice it
was not yet tested. So first integrators should expect rough edged.

The code is very basic and currently does not deal with huge amounts of data.
It just takes list of all available resource ids and process each id in a
synchronous way, one by one. This part should be improved, but if list of all
resource type ids fits into RAM, it should work fine.

Also qvarn-mr takes care of all synchronizations things automatically. That
means, when you deploy qvarn-mr for the first time, it will do initial
synchronization automatically, when new map/reduce handlers are added, they
will be updated automatically.


How to install
==============

The code was tested using python 3.4.5 and pip 9.0.1.

In order to install qvarnmr, create a virtualenv, activate it and run these
commands::

  pip install -f vendor -r requirements.txt -e . 

Yes this is install from source. Currently there is no any python package, nor
debian package. But python source package should be easily crated by running::

  python setup.py sdist

Then you can put this package to your project and install it from there.


How to run tests
================

Activate your virtualenv and run::

  py.test --cov-report=term-missing --cov=qvarnmr tests


How to use qvarnmr
==================

Firs of all, you have to define map/reduce handlers. You can read more about
how to define map/reduce handlers in the `How to define map/reduce handlers`_
section. Here is just a simple example:

.. code-block:: python

    from qvarnmr.func import join, item


    handlers = {
        'company_reports__map': {
            'orgs': {
                'type': 'map',
                'version': 1,
                'handler': item('id'),
            },
            'reports': {
                'type': 'map',
                'version': 1,
                'handler': item('org'),
            },
        },
        'company_reports': {
            'company_reports__map': {
                'type': 'reduce',
                'version': 1,
                'handler': join({
                    'org': {
                        'id': 'org_id',
                    },
                    'report': {
                        'id': 'report_id',
                    },
                }),
            },
        },
    }

Handlers should be defined in an importable Python file.

Then you have to define new derived Qvarn resource types in Qvarn resource
types yaml file. ``qvarn-mr`` also requires two resource tipes to manage
internal state:

.. code-block:: yaml

    path: /qvarnmr_listeners
    type: qvarnmr_listener
    versions:
    - prototype:
        id: ''
        type: ''
        revision: ''
        instance: ''
        resource_type: ''
        listener_id: ''
        owner: ''
        timestamp: ''
      version: v1

    path: /qvarnmr_handlers
    type: qvarnmr_handler
    versions:
    - prototype:
        id: ''
        type: ''
        revision: ''
        instance: ''
        target: ''
        source: ''
        version: 0
      version: v1

Also qvarnmr worker requires these Qvarn scopes::

    uapi_qvarnmr_listeners_get
    uapi_qvarnmr_listeners_post
    uapi_qvarnmr_listeners_id_get
    uapi_qvarnmr_listeners_id_put
    uapi_qvarnmr_listeners_id_delete
    uapi_qvarnmr_listeners_search_id_get

    uapi_qvarnmr_handlers_get
    uapi_qvarnmr_handlers_post
    uapi_qvarnmr_handlers_id_get
    uapi_qvarnmr_handlers_id_put
    uapi_qvarnmr_handlers_id_delete
    uapi_qvarnmr_handlers_search_id_get

    uapi_<target>_get
    uapi_<target>_post
    uapi_<target>_id_get
    uapi_<target>_id_put
    uapi_<target>_id_delete
    uapi_<target>_search_id_get

    uapi_<source>_get
    uapi_<source>_post
    uapi_<source>_id_get
    uapi_<source>_id_put
    uapi_<source>_id_delete
    uapi_<source>_search_id_get
    uapi_<source>_listeners_post
    uapi_<source>_listeners_id_get
    uapi_<source>_listeners_id_delete
    uapi_<source>_listeners_id_notifications_get
    uapi_<source>_listeners_id_notifications_id_get
    uapi_<source>_listeners_id_notifications_id_delete

Once you have defined handlers and new resource types, you run
``qvarnmr-worker``::

  qvarnmr-worker path.to.handlers -c path/to/qvarnmr.cfg -f

Here ``path.to.handlers`` is a Python path to your map/reduce handlers
configuration.

``-f`` stands for *forever*.

``-c path/to/qvarnmr.cfg`` is a Python configparser configuration file. Here is
example of configuration file:


.. code-block:: ini

    [qvarn]
    base_url = https://qvarn-example.tld,
    client_id = test_client_id
    client_secret = verysecret
    verify_requests = no
    scope = scope1,scope2,scope3

    [qvarnmr]
    instance = instance-name
    keep_alive_update_interval = 10  # seconds
    keep_alive_timeout = 60  # seconds

In this configuration file you need to specify connection parameters for the
Qvarn. Also you need to specify qvarnmr **instance name**. This name will be
used to know which notification handlers to use. There can be multiple qvarnmr
instances running, each processing different handlers. In order to distinguish
between these qvarnmr instances, instance name is used.

Probably it's a good idea to use project name as **instance name**. Because if
multiple projects will run on the same Qvarn database instance, then they will
not steal notifications from each other.

**keep_alive_update_interval** allows you to control how often a warker need
no announce that it is still alive and **keep_alive_timeout** is a time since
last update when a worker is considered as crashed in case if worker did not
had a chance to clean up befor exit. Even if a worker is crashed you will not
be able to run another until **keep_alive_timeout** time is passed. All times
are specified in seconds.

That's it.


How to define map/reduce handlers
=================================

Map/reduce handlers are defined in structure like this:

.. code-block:: python

    handlers = {
        'map-target': {
            'source-resource-type': {
                'type': 'map',
                'version': 1,
                'handler': a_maper_function,
            },
        },
        'reduce-target': {
            'map-target': {
                'type': 'reduce',
                'version': 1,
                'handler': a_maper_function,
            },
        },
    }

Here ``map-target`` and ``reduce-target`` are resource types, where derived
data from ``source-resource-type`` are stored. Client applications should not
write to derived resource types, because it might interfere with the qvarn-mr
engine. qvarn-mr expects that client applications only reads from derived
resources and only qvarn-mr takes care of populating data to the derived
resources.

There are some rules:

- You can have multiple sources for map targets.

- Reduce target can have only one source and that source must be a map target.


Let's analyse the following example a bit further:

.. code-block:: python

    from qvarnmr.func import join, item


    handlers = {
        'company_reports__map': {
            'orgs': {
                'type': 'map',
                'version': 1,
                'handler': item('id'),
            },
            'reports': {
                'type': 'map',
                'version': 1,
                'handler': item('org'),
            },
        },
        'company_reports': [
            'company_reports__map': {
                'type': 'reduce',
                'version': 1,
                'handler': join({
                    'org': {
                        'id': 'org_id',
                    },
                    'report': {
                        'id': 'report_id',
                    },
                }),
            },
        },
    }

Here we have two **source resource types** ``orgs`` and ``reports`` for the
**map handler**. Result of these map functions will be written to a new derived
resource type ``company_reports__map``, we will call these resource types as
**target resource types**.

Then everything, that goes  into ``company_reports__map`` will be processed by
the **reduce handler**, key by key. Result of the reduce handler will be
written into ``company_reports`` target resource type.

``item()`` and ``join()`` are helpers, to build a function for handling common
map/reduce tasks, like taking an item from source resource or joining multiple
resources into one. For such common tasks you don't need to define custom map
or reduce functions, you can use these helpers.

If there is no helper, you can always use your own functions.

How to define map function
--------------------------

Here is example of a map function:

.. code-block:: python

    def my_map_function(resource):
        return resource['id'], None


CHANGES.rst
Change History
==============

0.1.12 (unreleased)
-------------------

- Nothing changed yet.


0.1.11 (2018-05-02)
-------------------

- Avoid slow qvarn.search(..., show=("_mr_key",)) queries in resync that would
  time out and break resyncing for any larger resource tables.


0.1.10 (2018-04-19)
-------------------

- Log the number of notifications that require handling, and log progress
  messages after handling each batch of 100.

- Avoid loading resources from Qvarn when there are no registered handlers
  to use on those resources.

- Fix typo in log message ("can't find resource ... specifiend in notification"
  -> "... specified ...").


0.1.9 (2018-01-31)
------------------

- Fix bug when code if executed further on, even though exception was logged


0.1.8 (2018-01-25)
------------------

- Handle errors while getting Notification from qvarn

- Attempt to fix mystical listeners mutability bug.


0.1.7 (2017-10-09)
------------------

- Added proper handler of duplicates for reduced resources, where single key
  could have more than one resource. Now newest resource is used and all other
  duplicates re automatically removed.

  Client applications must add `_mr_timestamp` field for all reduce target
  resource types.

- Added strict handler configuration validator. After upgrading you might get
  various `HandlerValidationError` exceptions. Simply read the explanation in
  the error message and fix what is needed.

- `qvarnmr.testing.utils.process` now has `raise_errors` keyword argument which
  is `True` by default. Previously, `process` helper simply logged errors from
  handlers, but did not propagated any errors. Now behaviour changed and by
  default errors are propagated. When running tests usually you want to get all
  the errors loudly.

- Added retries if client app handler fails with error. Retries are tracked per
  notification. Single notification can be handled by multiple handers, but if
  at least one handler fails, all handlers will be retries even those, who
  completed successfully.

  When an exception comes from handler, then notification is left undeleted and
  information about error is stored in memory. Then first retry will be
  attempted after 0.25 seconds, second after 1.5 second and finally
  notification will be deleted and no more retires will be attempted.

  If `qvarnmr-worker` will be terminated, then all notifications will be
  retries, but retries can happen more than 2 times, because information about
  previous retries where stored in RAM memory.

  This mechanism is temporary, until parallel handler processing will be
  implemented.


0.1.6 (2017-09-26)
------------------

- Fix issue, when reduce handler errors where not properly handled during full
  resync.

- qvarnmr.processed.process_map and _process_reduce functions became private
  functions, while qvarnmr.processed.MapReduceEngine.process_reduce_handlers
  became public, bicause it is used by resyncer.

- Add logging for how many items where produced by map handler.


0.1.5 (2017-09-04)
------------------

- Fixed one source for multiple targets bug, when listeners gave resource
  conflict errors.

- Replaced deprecated ``logger.warn`` to ``logger.warning``.


0.1.4 (2017-08-31)
------------------

- Proper logging

- Configurable keep alive parameters

- Keep alive updates after each handler processing

- Fix reduce full resync bug


0.1.3 (unreleased)
------------------

- Now keep alive updates are done after each map/reduce handler processing.
  Previously keep alive updates happend only after each map/reduce changes
  batch.

- Keep alive update automatically refeshes state resource from Qvarn if time
  from alast update is bigger than timeout.

- process_changes function was refactored to MapReduceEngine.process_changes
  method.

- Proper logging was added, now it is possible to see what is happening inside
  qvarn-mr by looking at logs.

- Fix full reduce resync bug, it seems, that full reduce resync was blockin
  whole map/reduce process while resyncing.

- Configurable keep alive parameters using **keep_alive_update_interval** and
  **keep_alive_timeout** configuration file options.


0.1.2 (2017-07-14)
------------------

- Added checks to make sure that one worker is running. This required some
  schema changes, two fields ``owner`` and ``timestamp`` where added to
  ``qvarnmr_handlers`` resource type.


0.1.1 (2017-06-27)
------------------

- Reduce handlers now are processed using Qvarn notifications. This should
  solve possible cases related with eventual consistency.

- ``qvarnmr-resync`` script was replaced by automatic synchronisation. So now,
  it is enought to just run ``qvarnmr-worker`` and all the sinchronisation will
  happen automatically. You just need to restart ``qvarnmr-worker`` after
  updateing handlers configuration.


qvarnmr/processor.py
import time
import logging

from operator import itemgetter
from itertools import groupby
from collections import namedtuple

from qvarnmr.clients.qvarn import QvarnResourceNotFound
from qvarnmr.exceptions import HandlerVersionError
from qvarnmr.func import run
from qvarnmr.handlers import get_handlers
from qvarnmr.utils import is_empty

logger = logging.getLogger(__name__)


RESOURCE_CHANGES = CREATED, UPDATED, DELETED = ('created', 'updated', 'deleted')

Notification = namedtuple('Notification', [
    'resource_type',
    'resource_change',
    'resource_id',
    'notification_id',
    'listener_id',
    'generated',  # True - means notification is not coming from Qvarn, but is generated, and there
                  # is no point to delete it.
])

FailedNotification = namedtuple('FailedNotification', Notification._fields + (
    'retries',
    'processed_at',
))

Context = namedtuple('Context', [
    'qvarn',
    'source_resource_type',
])


def _same_version(version, resources):
    existing_versions = {x['_mr_version'] for x in resources}
    if len(resources) == 1 and version in existing_versions:
        return True
    else:
        return False


def _clean_existing_resources(qvarn, target_resource_type, resources):
    qvarn.delete_multiple(target_resource_type, [x['id'] for x in resources])


def _save_map_results(qvarn, handler, resource, target_resource_type, source_resource_type,
                      results):
    resources_updated = 0

    for key, value in results:
        if isinstance(value, dict):
            value['_mr_value'] = None
        else:
            value = {'_mr_value': value}

        value['_mr_key'] = key
        value['_mr_source_id'] = resource['id']
        value['_mr_source_type'] = source_resource_type
        value['_mr_deleted'] = False
        value['_mr_version'] = handler['version']

        qvarn.create(target_resource_type, value)
        resources_updated += 1

    return resources_updated


def _save_reduce_result(qvarn, handler, resource, target_resource_type, key, value):
    # If reduce function returns non-dict value, store it to _mr_value.
    if isinstance(value, dict):
        value['_mr_value'] = None
    else:
        value = {'_mr_value': value}

    # Also store _mr_key in order to be able to update this key with new
    # value, once one of the source resources will be changed.
    value['_mr_key'] = key

    # Save handler version to be able to track outdated resources.
    value['_mr_version'] = handler['version']

    # Timestamp in nanoseconds, Qvarn uses signed 8 bytes integer for storing numbers. So if we
    # store timestamp in nanoseconds, we have enough space until ~2270 year.
    value['_mr_timestamp'] = int(time.time() * 1e9)

    # Save reduced value to the target resource type.
    if resource is None:
        qvarn.create(target_resource_type, value)
    else:
        qvarn.update(target_resource_type, resource['id'], value)


def _process_map(qvarn, source_resource_type, resource_change, resource_id, handlers, resync=False):
    resources_updated = 0
    context = Context(qvarn, source_resource_type)
    if resource_change in (CREATED, UPDATED):
        resource = qvarn.get(source_resource_type, resource_id)
        for target_resource_type, handler in handlers:
            logger.info('processing map handler source=%s target=%s change=%s resource=%s '
                        'handler=%r version=%s resync=%r', source_resource_type,
                        target_resource_type, resource_change, resource_id, handler['handler'],
                        handler['version'], resync)
            start = time.time()

            existing_resources = qvarn.search(target_resource_type, _mr_source_id=resource['id'],
                                              show=('_mr_version',))

            if resync and _same_version(handler['version'], existing_resources):
                # If we are doning full resync, skip resources that are already resynced.
                continue

            # Run map handler.
            # If handler fails, nothing will be updated.
            results = list(run(handler['handler'], context, resource))

            # We have to clean all existing resources produced by map handler previously, because we
            # can't easily identify previously generated (key, value) pairs with the new ones.
            _clean_existing_resources(qvarn, target_resource_type, existing_resources)
            resources_updated += _save_map_results(qvarn, handler, resource, target_resource_type,
                                                   source_resource_type, results)
            logger.info('done processing map handler source=%s target=%s change=%s resource=%s '
                        'handler=%r version=%s resync=%r output=%d time=%.2fs',
                        source_resource_type, target_resource_type, resource_change, resource_id,
                        handler['handler'], handler['version'], resync, len(results),
                        time.time() - start)

    elif resource_change == DELETED:
        for target_resource_type, handler in handlers:
            logger.info('processing map handler source=%s target=%s change=%s resource=%s '
                        'handler=%r version=%s resync=%r', source_resource_type,
                        target_resource_type, resource_change, resource_id, handler['handler'],
                        handler['version'], resync)
            start = time.time()

            resources = qvarn.search(target_resource_type, _mr_source_id=resource_id)
            for resource in qvarn.get_multiple(target_resource_type, resources):
                # Until reduce was not yet processed, we don't want to delete this resource. Because
                # reduce handlers need to know the key.
                # All resources marked for deletion will be cleaned up after each update cycle.
                resource['_mr_deleted'] = True
                qvarn.update(target_resource_type, resource['id'], resource)
                resources_updated += 1

            logger.info('done processing map handler source=%s target=%s change=%s resource=%s '
                        'handler=%r version=%s resync=%r time=%.2fs', source_resource_type,
                        target_resource_type, resource_change, resource_id, handler['handler'],
                        handler['version'], resync, time.time() - start)

    else:
        raise ValueError('Unknown resource change type: %r' % resource_change)

    return resources_updated


def _map_reduce_resources(context, resources, handler):
    resources = context.qvarn.get_multiple(context.source_resource_type, resources)
    for resource in resources:
        for value in run(handler, context, resource):
            yield value


def _iter_reduce_resource_ids(qvarn, config, source_resource_type, key):
    resources = qvarn.search(source_resource_type, _mr_key=key,
                             show=('_mr_source_type', '_mr_version', '_mr_deleted'))
    for resource in resources:
        if not resource['_mr_deleted']:
            map_handler = config[source_resource_type][resource['_mr_source_type']]
            # We need to wait, while all source resources have same version and only then do the
            # reduce part, when all data are consistent.
            if map_handler['version'] != resource['_mr_version']:
                raise HandlerVersionError(key)
            yield resource['id']


def _get_and_ensure_single_resource(qvarn, resource_type, key):
    resources = qvarn.search(resource_type, _mr_key=key, show_all=True)

    if len(resources) > 1:
        resources = sorted(resources, key=lambda x: (x['_mr_timestamp'] or 0), reverse=True)
        _clean_existing_resources(qvarn, resource_type, resources[1:])

    if len(resources) > 0:
        return resources[0]


def _process_reduce(qvarn, config, source_resource_type, key, handlers, resync=False):
    context = Context(qvarn, source_resource_type)
    for target_resource_type, handler in handlers:
        logger.info('processing reduce handler source=%s target=%s key=%s handler=%r '
                    'version=%s resync=%r', source_resource_type, target_resource_type, key,
                    handler['handler'], handler['version'], resync)
        start = time.time()

        target_resource = _get_and_ensure_single_resource(qvarn, target_resource_type, key)

        if resync and target_resource and _same_version(handler['version'], [target_resource]):
            # If we are doning full resync, skip resources that are already resynced.
            continue

        # Get all resources by given key, force result to be an iterator,
        # because in future we should query resources iteratively in order to
        # avoid huge memory consumptions.
        resources = _iter_reduce_resource_ids(qvarn, config, source_resource_type, key)

        if 'map' in handler:
            resources = _map_reduce_resources(context, resources, handler['map'])

        resources, empty = is_empty(resources)
        if target_resource and empty:
            # Delete key entry if there are no keys produced by map handlers.
            _clean_existing_resources(qvarn, target_resource_type, [target_resource])

        else:
            # Call reduce function for all resources matching key.
            value = next(run(handler['handler'], context, resources), None)
            _save_reduce_result(qvarn, handler, target_resource, target_resource_type, key, value)

            logger.info('done processing reduce handler source=%s target=%s key=%s handler=%r '
                        'version=%s resync=%r time=%.2fs', source_resource_type,
                        target_resource_type, key, handler['handler'], handler['version'], resync,
                        time.time() - start)


class MapReduceEngine:
    EVENTS = (
        'map_handler_processed',
        'reduce_handler_processed',
    )

    def __init__(self, qvarn, config, raise_errors=False):
        self.qvarn = qvarn
        self.config = config
        self.raise_errors = raise_errors
        self.mappers, self.reducers = get_handlers(config)
        self.callbacks = {event: [] for event in self.EVENTS}
        self.reduce_handler_sources = {
            source
            for target, sources in config.items()
            for source, handler in sources.items()
            if handler['type'] == 'reduce'
        }

        self._failed_notifications = {}

    def _run_callbacks(self, event):
        for callback in self.callbacks[event]:
            callback()

    def _report_success(self, notifications):
        for notification in notifications:
            if notification.notification_id in self._failed_notifications:
                del self._failed_notifications[notification.notification_id]
            _delete_notification(self.qvarn, notification)

    def _report_error(self, notifications):
        for notification in notifications:
            key = notification.notification_id
            error = self._failed_notifications.get(key, None)
            if error is None:
                self._failed_notifications[key] = FailedNotification(**dict(
                    notification._asdict(),
                    retries=0,
                    processed_at=time.time(),
                ))
            else:
                if error.retries > 1:
                    del self._failed_notifications[key]
                    _delete_notification(self.qvarn, notification)
                else:
                    self._failed_notifications[key] = FailedNotification(**dict(
                        notification._asdict(),
                        retries=error.retries + 1,
                        processed_at=error.processed_at,
                    ))

    def _iter_changes(self, changes):
        for notification in changes:

            # Retry failed notifications.
            if notification.notification_id in self._failed_notifications:
                now = time.time()
                notification = self._failed_notifications[notification.notification_id]
                if notification.retries == 0 and now - notification.processed_at < 0.25:
                    logger.debug('retry 0.25 (skip): %s', now - notification.processed_at)
                    continue
                elif notification.retries == 1 and now - notification.processed_at < 1.5:
                    logger.debug('retry 1.5 (skip): %s', now - notification.processed_at)
                    continue
                elif notification.retries > 1:
                    logger.debug('retry > 1 (abort)')
                    del self._failed_notifications[notification.notification_id]
                    _delete_notification(self.qvarn, notification)
                    continue
                logger.debug("retrying failed notification, resource: %s id: %s, retry: %s "
                             "delay: %s", notification.resource_type, notification.resource_id,
                             notification.retries, now - notification.processed_at)

            yield notification

    def _process_map_handlers(self, changes, resync=False):
        changes_processed = 0
        errors = 0
        reduce_changes = []
        progress = 0

        # Run through all changes, process map handlers immediately and collect changes that have
        # reduce handlers for processing in groups in the next step.
        for notification in changes:
            try:
                handlers = self.mappers[notification.resource_type]
                if handlers:
                    _process_map(
                        self.qvarn, notification.resource_type, notification.resource_change,
                        notification.resource_id, handlers, resync,
                    )

            except Exception:
                # XXX: probably errors should be handler inside _process_map and another
                #      exception could be rerised with information about which handler failed.
                logger.exception("error while processing map handlers for %r", (
                    notification.resource_type, notification.resource_change,
                    notification.resource_id,
                ))
                self._report_error([notification])
                errors += 1
                if self.raise_errors:
                    raise

            else:
                should_reduce = (
                    notification.resource_type in self.reduce_handler_sources and

                    # We ignore all delete notifications, since we don't delete mapped resources, we
                    # mark then as deleted first (and that generates UPDATED notification) and only
                    # then mapped resources are deleted (cleaned) completely. And once they are
                    # deleted for real, we are no longer interested in them.
                    notification.resource_change != DELETED
                )
                if should_reduce:
                    resource = self.qvarn.search_one(notification.resource_type,
                                                     id=notification.resource_id,
                                                     show=('_mr_key',), default=None)
                    if resource is None:
                        logger.warning(
                            "can't find resource (%s, %s) specified in notificaton, the resource "
                            "could be deleted or not yet replicated", notification.resource_type,
                            notification.resource_id)
                        self._report_error([notification])
                        errors += 1
                    else:
                        # Collect all changes that have reduce handlers and process them later,
                        # grouped by key. This will lower number of reduce handler calls.
                        reduce_changes.append(((notification.resource_type, resource['_mr_key']),
                                               notification))
                else:
                    self._report_success([notification])
                    changes_processed += 1

            self._run_callbacks('map_handler_processed')

            progress += 1
            if progress % 100 == 0:
                logger.info('processed %d notifications', progress)

        return changes_processed, errors, reduce_changes

    def process_reduce_handlers(self, changes, *, errors=0, resync=False):
        changes_processed = 0
        progress = 0

        # Process all changes with reduce handlers in groups.
        changes = sorted(changes, key=itemgetter(0))
        grouped = [(key, list(group)) for key, group in groupby(changes, key=itemgetter(0))]
        if grouped:
            logger.info("grouped %d changes into %d groups", len(changes), len(grouped))
        for (source_resource_type, key), group in grouped:
            try:
                _process_reduce(self.qvarn, self.config, source_resource_type, key,
                                self.reducers[source_resource_type], resync=resync)

            except HandlerVersionError as e:
                # If we end up here, it means, that this key has inconsistent versions in mapped
                # resources. In that case we postpone notification by leaving undeleted.
                logger.debug("incompatible mapped resource versions for key=%r of %r resource.",
                             e.key, source_resource_type)
                self._report_error([notification for _, notification in group])

            except Exception:
                # XXX: probably errors should be handler inside _process_reduce and another
                #      exception could be rerised with information about which handler failed.
                logger.exception("error while processing reduce handlers for %r, key=%r",
                                 source_resource_type, key)
                notifications = [notification for _, notification in group]
                self._report_error(notifications)
                errors += len(notifications)
                if self.raise_errors:
                    raise

            else:
                # Delete processed mapped resources if they where marked for deletion.
                for resource_id in self.qvarn.search(source_resource_type, _mr_key=key,
                                                     _mr_deleted=True):
                    self.qvarn.delete(source_resource_type, resource_id)

                notifications = [notification for _, notification in group]
                self._report_success(notifications)
                changes_processed += len(notifications)

            progress += 1
            if progress % 100 == 0:
                logger.info('processed %d/%d reduce handlers', progress, len(grouped))

            self._run_callbacks('reduce_handler_processed')

        return changes_processed, errors

    def add_callback(self, event, callback):
        self.callbacks[event].append(callback)

    def process_changes(self, changes, resync=False):
        logger.info('processing changes resync=%r', resync)
        start = time.time()
        changes = self._iter_changes(changes)
        mapped, errors, reduce_changes = self._process_map_handlers(changes, resync)
        reduced, errors = self.process_reduce_handlers(reduce_changes, errors=errors, resync=resync)
        logger.info('done processing changes resync=%r mapped=%d reduced=%d errors=%d '
                    'time=%.2fs', resync, mapped, reduced, errors, time.time() - start)
        return mapped + reduced


def get_changes(qvarn, listeners):
    l = list(listeners)  # create a new copy of listeners
    for resource_type, listener, state in l:
        path = resource_type + '/listeners/' + listener['id'] + '/notifications'
        notifications = qvarn.get_list(path)
        if notifications:
            logger.info("there are %d pending notifications for source=%s",
                        len(notifications), resource_type)
        for notification_id in notifications:
            try:
                notification = qvarn.get(path, notification_id)
            except QvarnResourceNotFound:
                logger.warning("master:   notification has been deleted (probably after giving up "
                               "retries): notification=%s resource_type=%s", notification_id,
                               resource_type)
                # continue loop withouth yielding
                continue
            yield Notification(
                resource_type=resource_type,
                resource_change=notification['resource_change'],
                resource_id=notification['resource_id'],
                notification_id=notification['id'],
                listener_id=listener['id'],
                generated=False,
            )


def _delete_notification(qvarn, notification):
    if not notification.generated:
        logger.debug("delete notification for resource type=%s change=%s resource=%s",
                     notification.resource_type, notification.resource_change,
                     notification.resource_id)

        path = notification.resource_type + '/listeners/' + notification.listener_id + '/notifications'
        qvarn.delete(path, notification.notification_id)
